### PR TITLE
Adding DynamoDB High Availability Features

### PR DESCRIPTION
**Pull request by kmg-stripe**
### Context

This is a forked version of the branch for [Adding DynamoDB high availability features](https://github.com/Netflix/mantis/pull/675)

The motivation for this PR is to ensure we can build a snapshot using the original branch, which was failing since the branch is in a fork of Mantis.  Context is here: https://github.com/Netflix/mantis/pull/675#issuecomment-2173895730 

### Checklist

- [X] `./gradlew build` compiles code correctly
- [X] Added new tests where applicable
- [X] `./gradlew test` passes all tests
- [X] Extended README or added javadocs where applicable

### CODE REVIEW

**Comment by kmg-stripe** [comment]
@Andyz26 - I pulled Mike's branch into my forked repo, pushed the branch and created this PR to test building a snapshot.  It looks like I need an approval to start the build.

**Comment by Andyz26** [comment]
> @Andyz26 - I pulled @mcowgill-stripe's branch into my forked repo, pushed the branch and created this PR to test building a snapshot. It looks like I need an approval to start the build.

what did you see when you click the details/approval link on the "publish snapshot to ..." task above? I thought the collab/write permission groups give you permission to approve there.
BTW I've approved the build.

**Comment by kmg-stripe** [comment]
> > @Andyz26 - I pulled @mcowgill-stripe's branch into my forked repo, pushed the branch and created this PR to test building a snapshot. It looks like I need an approval to start the build.
> 
> what did you see when you click the details/approval link on the "publish snapshot to ..." task above? I thought the collab/write permission groups give you permission to approve there. BTW I've approved the build.

@Andyz26 - Ah, I do not see anything there right now, so assume I missed it.  I'll self-approve next time, assuming I see the approval link.

It looks like the build finished successfully.  Based on the comments in #675 , it looks like you wanted to do some testing on the snapshots before we merge it?  I also see some comments in #675  .  Are you ok with addressing those in follow-on PR(s)?

**Comment by Andyz26** [comment]
> > > @Andyz26 - I pulled @mcowgill-stripe's branch into my forked repo, pushed the branch and created this PR to test building a snapshot. It looks like I need an approval to start the build.
> > 
> > 
> > what did you see when you click the details/approval link on the "publish snapshot to ..." task above? I thought the collab/write permission groups give you permission to approve there. BTW I've approved the build.
> 
> @Andyz26 - Ah, I do not see anything there right now, so assume I missed it. I'll self-approve next time, assuming I see the approval link.
> 
> It looks like the build finished successfully. Based on the comments in #675 , it looks like you wanted to do some testing on the snapshots before we merge it? I also see some comments in #675 . Are you ok with addressing those in follow-on PR(s)?

yeah I need to try this build to see if anything is broken. And yeah for major refactor comments/discusion thread they could defer to follow up PRs.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java [hunk 1]
```diff
@@ -346,20 +288,6 @@ public MasterConfiguration getConfig() {
...
 
-    public String getDescriptionJson() {
-        try {
-            return DefaultObjectMapper.getInstance().writeValueAsString(leadershipManager.getDescription());
-        } catch (JsonProcessingException e) {
-            throw new IllegalStateException(String.format("Failed to convert the description %s to JSON: %s", leadershipManager.getDescription(), e.getMessage()), e);
-        }
-    }
-
-    public Observable<MasterDescription> getMasterObservable() {
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:357]: looks like we still need to expose this from the new interface as we need it for some internal connections.

Comment (Andyz26) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:357]: "Observable<MasterDescription>" from the leader client.

Comment (kmg-stripe) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:357]: Just to be clear, do you mean expose this from `ILeadershipManager` or another interface?

Comment (Andyz26) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:357]: we still need "public Observable<MasterDescription> getMasterObservable()" method from teh masterMain class.

Comment (kmg-stripe) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:357]: ah, understood!  i'll push up a commit later today.

Comment (kmg-stripe) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:357]: @Andyz26 I pushed up a commit a few hours ago.

Also, I do not see anyway for me to approve snapshot creation.  I just see: "This branch is waiting to be deployed" with "Integrate Pull Request" with a link to the build.

Comment (Andyz26) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:357]: just approved. I guess there is another security setting somewhere I need to find.

Comment (kmg-stripe) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:357]: hey @Andyz26 - Just checking in to see if you found any more issues :)

Comment (Andyz26) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:357]: I got the control plane tested with some config changes and it seems to be working now. I will tried the agent/runtime parts today.

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/CoreConfiguration.java [hunk 2]
```diff
@@ -77,4 +77,8 @@ public interface CoreConfiguration {
     @Config("mantis.asyncHttpClient.readTimeoutMs")
     @Default("10000")
     int getAsyncHttpClientReadTimeoutMs();
+
+    @Config("mantis.leader.monitor.factory")
+    @Default("io.mantisrx.server.core.master.LocalLeaderFactory")
+    String getLeaderMonitorFactory();
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/CoreConfiguration.java:83]: nit: getLeaderMonitorFactoryName

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/ILeadershipManager.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.mantisrx.server.core;
+/*
+ * Copyright 2019 Netflix, Inc.
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/ILeadershipManager.java:18]: dupe copyright?

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/master/ZookeeperMasterMonitor.java [hunk 4]
```diff
@@ -36,36 +37,45 @@
...
-        this.nodeMonitor = new NodeCache(curator, masterPath);
-        this.jsonSerializer = new JsonSerializer();
+        this.nodeMonitor = new NodeCache(curator.getCurator(), masterPath);
     }
 
     @Override
-    public void startUp() throws Exception {
+    public void start()  {
+        try {
+            if(curator.getCurator().getState() != CuratorFrameworkState.STARTED) {
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/master/ZookeeperMasterMonitor.java:60]: just checking: is this thread safe?

File: mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBLeaderElector.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DynamoDBLeaderElector extends BaseService {
+
+    private static final Logger logger = LoggerFactory.getLogger(DynamoDBLeaderElector.class);
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBLeaderElector.java:41]: nit: use "@Slf4j"

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java [hunk 6]
```diff
@@ -90,10 +83,11 @@ public void stateChanged(CuratorFramework client, ConnectionState newState) {
     @Override
     public void start() {
         try {
-            isConnectedGauge.set(0L);
-            setupCuratorListener();
-            curator.start();
-            masterMonitor.startAsync().awaitRunning();
+            if(curator.getState() != CuratorFrameworkState.STARTED) {
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java:86]: nit: add logging

**Review by Andyz26 — APPROVED** [review]

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/master/ZookeeperMasterMonitor.java (hunk 4 above)
Comment (kmg-stripe) [reply mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/master/ZookeeperMasterMonitor.java:60]: ah, yeah, looks like it is not.  it does seem that there is the potential for racing "starts" to result in a thrown exception here.

i think a small refactor can fix this.  i'll do that today.

Comment (kmg-stripe) [reply mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/master/ZookeeperMasterMonitor.java:60]: oh wait, i take that back, concurrent calls to "start()" should be safe.  subsequent calls should be no-ops.

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java [hunk 7]
```diff
@@ -90,10 +83,11 @@ public void stateChanged(CuratorFramework client, ConnectionState newState) {
     @Override
     public void start() {
         try {
-            isConnectedGauge.set(0L);
-            setupCuratorListener();
-            curator.start();
-            masterMonitor.startAsync().awaitRunning();
```
Comment (kmg-stripe) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java:96]: @Andyz26 - re: thread safety: wdyt of wrapping this in `synchronized (this) { ... }` to synchronize concurrent calls to `start()`?

File: mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBStore.java [hunk 8]
```diff
@@ -115,7 +115,7 @@ public Map<String, String> getAll(String tableName, String partitionKey) throws
         expressionAttributesNames.put(SK_E, SK);
         Map<String, AttributeValue> expressionAttributeValues = new HashMap<>();
         expressionAttributeValues.put(PK_V, AttributeValue.builder().s(tableName).build());
-        expressionAttributeValues.put(SK_V, AttributeValue.builder().s(partitionKey).build());
+        expressionAttributeValues.put(SK_V, AttributeValue.builder().s(String.format("%s#", partitionKey)).build());
```
Comment (kmg-stripe) [inline mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBStore.java:118]: This is unrelated to the refactor, but a bug found while testing

File: mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java [hunk 9]
```diff
@@ -83,10 +83,13 @@ public void stateChanged(CuratorFramework client, ConnectionState newState) {
     @Override
     public void start() {
         try {
-            if(curator.getState() != CuratorFrameworkState.STARTED) {
-                isConnectedGauge.set(0L);
-                setupCuratorListener();
-                curator.start();
+            synchronized (this) {
```
Comment (kmg-stripe) [inline mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/zookeeper/CuratorService.java:86]: @Andyz26 I think this should handle the potential race when start() can be called more than once (your comment on thread safety).  lmk if you think this is not correct, or can think of a better way...

**Comment by kmg-stripe** [comment]
@Andyz26 - i addressed your comments and fixed a small, unrelated bug in the dynamodb store.  lmk if there is anything else.  otherwise, i'll merge :)

**Pull request merged**